platform/security/idp/security-idp-plugin/security-idp-plugin-nameidpolicy/src/main/java/org/codice/ddf/security/idp/plugin/nameidpolicy/NameIdPolicyPresignPlugin.java

```diff
@@ -0,0 +1,154 @@
+import ddf.security.SubjectUtils;
+import ddf.security.samlp.SamlProtocol.Binding;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Predicate;
+import java.util.stream.Stream;
+import javax.annotation.Nullable;
+import org.apache.commons.lang.StringUtils;
+import org.codice.ddf.security.idp.plugin.SamlPresignPlugin;
+import org.opensaml.core.xml.schema.XSString;
+import org.opensaml.saml.saml2.core.Attribute;
+import org.opensaml.saml.saml2.core.AuthnRequest;
+import org.opensaml.saml.saml2.core.NameID;
+import org.opensaml.saml.saml2.core.NameIDPolicy;
+import org.opensaml.saml.saml2.core.Response;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This plugin is responsible for setting the <@code>NameID</@code>'s <@code>Format</@code> &
+ * <@code>SPNameQualifier</@code> on the outgoing <@code>Response</@code>. It does this by parsing
+ * the <@code>NameIDPolicy</@code> of the passed in <@code>AuthnRequest</@code>.
+ *
+ * <p>See section 3.4.1.1 of the SAML Core specification for more information.
+ */
+public class NameIdPolicyPresignPlugin implements SamlPresignPlugin {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(NameIdPolicyPresignPlugin.class);
+
+  @Override
+  public void processPresign(
+      Response response,
+      AuthnRequest authnRequest,
+      List<String> spMetadata,
+      Set<Binding> supportedBindings) {
+    NameIDPolicy nameIdPolicy = authnRequest.getNameIDPolicy();
+
+    if (nameIdPolicy == null
+        || (StringUtils.isEmpty(nameIdPolicy.getFormat())
+            && StringUtils.isEmpty(nameIdPolicy.getSPNameQualifier()))) {
+      return;
+    }
+
+    response
+        .getAssertions()
+        .stream()
+        .map(assertion -> assertion.getSubject().getNameID())
+        .forEach(
+            assertionNameId ->
+                setAssertionNameIdQualifierAndFormat(
+                    assertionNameId,
+                    nameIdPolicy.getFormat(),
+                    nameIdPolicy.getSPNameQualifier(),
+                    response));
+  }
+
+  private void setAssertionNameIdQualifierAndFormat(
+      NameID assertionNameId,
+      @Nullable String nameIdFormatPolicy,
+      @Nullable String spNameQualifierPolicy,
+      Response response) {
+
+    if (StringUtils.isNotEmpty(spNameQualifierPolicy)) {
+      assertionNameId.setSPNameQualifier(spNameQualifierPolicy);
+    }
+
+    // the Assertion's NameID Format is set to the NameIDPolicy Format after the switch statement
+    if (StringUtils.isNotEmpty(nameIdFormatPolicy)) {
+      switch (nameIdFormatPolicy) {
+          // supported NameIDFormats
+        case NameID.UNSPECIFIED:
+          return; // avoid changing the Format later
+        case NameID.PERSISTENT:
+          // TODO DDF-3965
+          break;
+
+          // partially supported NameIDFormats
+        case NameID.X509_SUBJECT:
+          LOGGER.warn("Specifying the \"X509Subject\" NameIDPolicy Format is not fully supported.");
```

Comment:
❓ Should we log anything to the security log here? Same for the others.

Reply:
I'm not sure what our security logging practices are. This may be a good spot for it.

@blen-desta , @brjeter , @coyotesqrl ?

Reply:
I don't _think_ it's important enough to audit, as it's not a 👍/ 👎 decision and no rights are granted or denied because of this.